### ISSUE TITLE
What is fastText's cost function?

### ISSUE THREAD

**Issue by slederman** [facebookresearch/fastText#839]
Is FastText using cross-entropy as its cost function? If not, what is its cost function? In either case, what is the algorithm? Feel free to point me to a line of the source code.

**Comment by Celebio** [comment]
Hi @slederman ,
Several loss functions are available in fastText :
- softmax : this is basically [`-log` on the target](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L343) after applying the softmax function on the output layer
- hierarchical softmax  (as explained [here](https://fasttext.cc/docs/en/supervised-tutorial.html#advanced-readers-hierarchical-softmax)) which uses cross-entropy loss [on each node of the tree](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L258)
- negative sampling, that uses cross-entropy on [one true label](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L173) and on [some randomly chosen](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L177) negative labels
- one-vs-all which applies [cross-entropy on each label](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L139) independently.

You can find the computation of the cross-entropy inside `loss.cc` (it is named binaryLogistic) [here](https://github.com/facebookresearch/fastText/blob/b5b7d307274ce00ef52198fbc692ed3bd11d9856/src/loss.cc#L98).
Best regards,
Onur
